docs/setup/independent/high-availability.md

```diff
@@ -394,6 +402,82 @@ The next step is to create a Load Balancer that sits in front of your master nod
 
 If possible, use a smart load balancing algorithm like "least connections", and use health checks so unhealthy nodes can be removed from circulation. Most providers will provide these features.
+{% endcapture %}
+
+{% capture onsite %}
+In an on-site environment there may not be a physical load balancer available. Instead, keepalived can be used to setup a virtual IP pointing to a healthy master node. The configuration shown here provides an _active/passive_ setup rather than _real_ load balancing, but it can be extended for this purpose quite easily by setting up HAProxy, nginx or similar on the master nodes (not covered here). 
+
+1. Install keepalived, e.g. using your distribution's package manager. The configuration shown here works with version 1.3.5 and supposedly many others. Make sure to have it enabled (chkconfig, systemd, ...) so that it starts automatically when the respective node comes up.
+
+2. Create the following configuration file _/etc/keepalived/keepalived.conf_ on all master nodes:
+
+    ```shell
+    ! Configuration File for keepalived
+    global_defs {
+      router_id LVS_DEVEL
+    }
+    
+    vrrp_script check_apiserver {
+      script "/etc/keepalived/check_apiserver.sh"
+      interval 3
+      weight -2
+      fall 10
+      rise 2
+    }
+    
+    vrrp_instance VI_1 {
+        state <STATE>
+        interface <INTERFACE>
+        virtual_router_id 51
+        priority <PRIORITY>
+        authentication {
+            auth_type PASS
+            auth_pass 4be37dc3b4c90194d1600c483e10ad1d
+        }
+        virtual_ipaddress {
+            <VIRTUAL-IP>
+        }
+        track_script {
+            check_apiserver
+        }
+    }
+    ```
+
+    In the section `vrrp_instance VI_1`, change few lines depending on your setup:
+
+    * `state` is either `MASTER` (on the first master nodes) or `BACKUP` (the other master nodes).
+    * `interface` is the name of an existing public interface to bind the virtual IP to (usually the primary interface).
+    * `priority` should be higher for the first master node, e.g. 101, and lower for the others, e.g. 100.
+    * `auth_pass` use any random string here.
+    * `virtual_ipaddresses` should contain the virtual IP for the master nodes.
+
+3. Install the following health check script to _/etc/keepalived/check_apiserver.sh_ on all master nodes:
+
+    ```shell
+    #!/bin/sh
+
+    errorExit() {
+        echo "*** $*" 1>&2
+        exit 1
+    }
+
+    curl --silent --max-time 2 --insecure https://localhost:6443/ -o /dev/null || errorExit "Error GET https://localhost:6443/"
+    if ip addr | grep -q <VIRTUAL-IP>; then
+        curl --silent --max-time 2 --insecure https://<VIRTUAL-IP>:6443/ -o /dev/null || errorExit "Error GET https://<VIRTUAL-IP>:6443/"
```

Comment:
It might be better to use the /healthz or /healthz/ready endpoints here.

Reply:
Unfortunately it has turned out that an anonymous user cannot get path _/healthz_ or _/healthz/ready_. I don't think I'd want to switch authentication off on the API server. Hence I'll have to stick with a simple http check here. 

Would it be a good idea to change the API server to allow anonymous access to the health check URLs?

Reply:
Odd, https://github.com/kubernetes/kubernetes/issues/51076 seems to indicate that /healthz should be available to unauthenticated users with the default RBAC configuration. Are we overriding the default config in kubeadm for a particular reason?

Reply:
It's the configuration created by _kubeadm_. I did not change this setting.

Reply:
I just tested this on a cluster created w/ the [Heptio Quickstart](https://aws.amazon.com/quickstart/architecture/heptio-kubernetes/), which uses kubeadm and I am able to hit the the /healthz endpoint without authentication
```
curl -k https://heptio-ku-apiloadb-mygytr5iaq5p-970181078.us-east-1.elb.amazonaws.com/healthz
ok
```

The kubeadm-config ConfigMap for the cluster is:
```
apiVersion: v1
data:
  MasterConfiguration: |
    api:
      advertiseAddress: 10.0.20.246
      bindPort: 6443
    apiServerCertSANs:
    - heptio-ku-apiloadb-mygytr5iaq5p-970181078.us-east-1.elb.amazonaws.com
    authorizationModes:
    - Node
    - RBAC
    certificatesDir: /etc/kubernetes/pki
    cloudProvider: aws
    etcd:
      caFile: ""
      certFile: ""
      dataDir: /var/lib/etcd
      endpoints: null
      image: ""
      keyFile: ""
    imageRepository: gcr.io/google_containers
    kubeProxy:
      config:
        bindAddress: 0.0.0.0
        clientConnection:
          acceptContentTypes: ""
          burst: 10
          contentType: application/vnd.kubernetes.protobuf
          kubeconfig: /var/lib/kube-proxy/kubeconfig.conf
          qps: 5
        clusterCIDR: ""
        configSyncPeriod: 15m0s
        conntrack:
          max: null
          maxPerCore: 32768
          min: 131072
          tcpCloseWaitTimeout: 1h0m0s
          tcpEstablishedTimeout: 24h0m0s
        enableProfiling: false
        featureGates: ""
        healthzBindAddress: 0.0.0.0:10256
        hostnameOverride: ""
        iptables:
          masqueradeAll: false
          masqueradeBit: 14
          minSyncPeriod: 0s
          syncPeriod: 30s
        ipvs:
          minSyncPeriod: 0s
          scheduler: ""
          syncPeriod: 30s
        metricsBindAddress: 127.0.0.1:10249
        mode: ""
        oomScoreAdj: -999
        portRange: ""
        resourceContainer: /kube-proxy
        udpTimeoutMilliseconds: 250ms
    kubeletConfiguration: {}
    kubernetesVersion: v1.9.2
    networking:
      dnsDomain: cluster.local
      podSubnet: ""
      serviceSubnet: 10.96.0.0/12
    nodeName: ip-10-0-20-246.ec2.internal
    token: ""
    tokenTTL: 0s
    unifiedControlPlaneImage: ""
kind: ConfigMap
metadata:
  creationTimestamp: 2018-02-22T17:07:18Z
  name: kubeadm-config
  namespace: kube-system
  resourceVersion: "141"
  selfLink: /api/v1/namespaces/kube-system/configmaps/kubeadm-config
  uid: d6856eb3-17f2-11e8-8121-12ecccfcf364
```

Reply:
Looks pretty much the same as on my system. But I am running on-prem instead on aws of course.